File: src/platform/llm/implementations/anthropic.ts
import { LLMProvider, GenerateRequest, GenerateResponse, LLMProviderCapability } from '../types';
import { ProviderError, RateLimitError, AuthenticationError, InvalidRequestError } from '../../../core/errors/llm-errors';
import { Logger } from '../../../core/telemetry';

interface AnthropicMessage {
  role: 'user' | 'assistant';
  content: string;
}

interface AnthropicRequest {
  model: string;
  messages: AnthropicMessage[];
  system?: string;
  max_tokens: number;
  temperature?: number;
  stop_sequences?: string[];
}

interface AnthropicContentBlock {
  type: string;
  text?: string;
}

interface AnthropicResponse {
  content?: AnthropicContentBlock[];
  usage?: {
    input_tokens: number;
    output_tokens: number;
  };
  stop_reason?: string;
}

interface AnthropicErrorResponse {
  error?: {
    message: string;
    type?: string;
  };
}

export class AnthropicProvider implements LLMProvider {
  readonly providerId = 'anthropic';
  readonly capabilities: LLMProviderCapability = {
    capabilities: new Set(['json_mode', 'streaming', 'vision']),
    maxTokens: 4096,
    contextWindow: 200000,
  };

  constructor(private config: { apiKey: string; model: string }) {}

  async generate(request: GenerateRequest): Promise<GenerateResponse> {
    try {
      if (!this.config.apiKey) throw new Error('Missing API Key');

      const body: AnthropicRequest = {
        model: this.config.model,
        messages: [{ role: 'user', content: request.prompt }],
        max_tokens: request.maxTokens ?? this.capabilities.maxTokens,
        temperature: request.temperature ?? 0.7,
      };

      if (request.systemPrompt) {
        body.system = request.systemPrompt;
      }

      if (request.stopSequences && request.stopSequences.length > 0) {
        body.stop_sequences = request.stopSequences;
      }

      Logger.debugProvider(this.providerId, 'REQUEST', body);

      const response = await fetch('https://api.anthropic.com/v1/messages', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': this.config.apiKey,
          'anthropic-version': '2023-06-01',
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({})) as AnthropicErrorResponse;
        const errorMessage = errorData.error?.message || response.statusText;
        throw new Error(`${response.status}: ${errorMessage}`);
      }

      const data: AnthropicResponse = await response.json();
      Logger.debugProvider(this.providerId, 'RESPONSE', data);

      const text = data.content
        ?.filter((block) => block.type === 'text')
        .map((block) => block.text || '')
        .join('\n')
        .trim() || '';

      const usage = data.usage || { input_tokens: 0, output_tokens: 0 };

      return {
        text,
        usage: {
          promptTokens: usage.input_tokens,
          completionTokens: usage.output_tokens,
          totalTokens: usage.input_tokens + usage.output_tokens,
        },
        finishReason: this.mapFinishReason(data.stop_reason ?? 'unknown'),
      };
    } catch (e: unknown) {
      throw this.normalizeError(e);
    }
  }

  private mapFinishReason(reason: string): GenerateResponse['finishReason'] {
    switch (reason) {
      case 'end_turn':
      case 'stop_sequence': return 'stop';
      case 'max_tokens': return 'length';
      default: return 'unknown';
    }
  }

  private normalizeError(e: unknown): Error {
    const message = e instanceof Error ? e.message : String(e);
    if (message.includes('429')) return new RateLimitError(message);
    if (message.includes('401') || message.includes('403')) return new AuthenticationError(message);
    if (message.includes('400')) return new InvalidRequestError(message);
    return new ProviderError(message);
  }
}

File: crates/wfs-core/src/llm/anthropic.rs
use async_trait::async_trait;
use reqwest::Client;
use std::collections::HashSet;

use crate::error::LlmError;
use crate::types::{FinishReason, GenerateRequest, GenerateResponse, ModelCapability, ProviderCapabilities, Usage};
use super::LlmProvider;

pub struct AnthropicProvider {
    api_key: String,
    model: String,
    client: Client,
    caps: ProviderCapabilities,
}

impl AnthropicProvider {
    pub fn new(api_key: String, model: String) -> Self {
        Self {
            api_key,
            model,
            client: Client::new(),
            caps: ProviderCapabilities {
                capabilities: HashSet::from([
                    ModelCapability::JsonMode,
                    ModelCapability::Streaming,
                    ModelCapability::Vision,
                ]),
                max_tokens: 4096,
                context_window: 200000,
            },
        }
    }
}

#[async_trait]
impl LlmProvider for AnthropicProvider {
    fn provider_id(&self) -> &'static str {
        "anthropic"
    }

    fn capabilities(&self) -> &ProviderCapabilities {
        &self.caps
    }

    async fn generate(&self, request: &GenerateRequest) -> Result<GenerateResponse, LlmError> {
        let mut body = serde_json::json!({
            "model": self.model,
            "messages": [{"role": "user", "content": &request.prompt}],
            "max_tokens": request.max_tokens.unwrap_or(self.caps.max_tokens),
            "temperature": request.temperature.unwrap_or(0.7),
        });

        if let Some(ref system) = request.system_prompt {
            body["system"] = serde_json::json!(system);
        }

        if let Some(ref stop) = request.stop_sequences {
            if !stop.is_empty() {
                body["stop_sequences"] = serde_json::json!(stop);
            }
        }

        let response = self
            .client
            .post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .json(&body)
            .send()
            .await
            .map_err(|e| LlmError::Provider(format!("Request failed: {e}")))?;

        let status = response.status();
        if !status.is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(normalize_error(&format!("{status}: {error_text}")));
        }

        let data: serde_json::Value = response
            .json()
            .await
            .map_err(|e| LlmError::Parse(format!("Failed to parse response: {e}")))?;

        let text = data["content"]
            .as_array()
            .map(|blocks| {
                blocks
                    .iter()
                    .filter(|b| b["type"] == "text")
                    .filter_map(|b| b["text"].as_str())
                    .collect::<Vec<_>>()
                    .join("\n")
            })
            .unwrap_or_default();

        let input_tokens = data["usage"]["input_tokens"].as_u64().unwrap_or(0) as u32;
        let output_tokens = data["usage"]["output_tokens"].as_u64().unwrap_or(0) as u32;

        let usage = Usage {
            prompt_tokens: input_tokens,
            completion_tokens: output_tokens,
            total_tokens: input_tokens + output_tokens,
        };

        let finish_reason = match data["stop_reason"].as_str() {
            Some("end_turn" | "stop_sequence") => FinishReason::Stop,
            Some("max_tokens") => FinishReason::Length,
            _ => FinishReason::Unknown,
        };

        Ok(GenerateResponse { text, usage, finish_reason })
    }
}

fn normalize_error(msg: &str) -> LlmError {
    if msg.contains("429") || msg.to_lowercase().contains("rate limit") {
        LlmError::RateLimit(msg.to_string())
    } else if msg.contains("401") || msg.contains("403") {
        LlmError::Authentication(msg.to_string())
    } else if msg.contains("400") {
        LlmError::InvalidRequest(msg.to_string())
    } else if msg.contains("timeout") || msg.to_lowercase().contains("timed out") {
        LlmError::Timeout(msg.to_string())
    } else {
        LlmError::Provider(msg.to_string())
    }
}
